### ISSUE TITLE
[GJEP] Design of a `Recurse` combinator

### ISSUE THREAD

**Issue by femtomc** [genjax-community/genjax#234]
This issue explores the design and implementation of a `Recurse` combinator in GenJAX. The goal is to assess what design DOFs are available for `genjax.Recurse`, understanding the limitations compared to `Gen.jl`.

## Background

To understand this issue, it's likely useful to understand the `Recurse` implementation in `Gen.jl`, as well as Marco's description of `Recurse` in his thesis. Here's the relevant blurb from Marco's thesis:

<img width="756" alt="image" src="https://user-images.githubusercontent.com/34410036/219404293-0a915ade-9f51-4d32-8bed-ed21a5ce88ec.png">

**Miscellaneous note on copy/sharing in Gen.jl**
> One thing I noticed is that the incremental gen.Recurse implementation copies the input data xs for each production (see: https://github.com/probcomp/GenExamples.jl/blob/6bc4a840fe6d823e008c1c4bf69e67ae1076e18e/gp_structure/incremental.jl#L19-L31), which is not necessary for the example — but is necessary if you want the first dimension of the xs in the recursion (e.g. see diagram https://www.mct.dev/assets/mct-thesis.pdf on 177) to indicate when to halt the recursion.

## Design

Here, I'm considering a design which slightly deviates from Gen.jl, which captures the pattern illustrated above:

* Fix input auxiliary arguments `xs` — these are not copied in each `production_kernel` call.

* Generate a “likelihood tree” template from a grammar. This process is characterized by 2 static constants, and 1 dynamic value. 

The 2 static constants are a maximum depth bound, and the maximum branch factor of the grammar. The dynamic value is a dynamic-allowed depth limit (similar to the sequence length value for Unfold).

* For production, produce enumeration indices in an array with size `maximum_branch_factor` — when a grammar production produces something below this factor, we pad and add a boolean mask (which we can take care of in the implementation of recurse).

* The result of production is a masked array of enumeration values — a “grammar index” which uniquely describes a sampled expression from the grammar. It has linear size:

$$\sum_{i=1}^{\text{dynamic depth limit}} (\text{max branch factor}) ^ i$$

The nice thing is that, in recurse, we can track the parent of each “production slice” (e.g. give me parent, I can get children generated by production_kernel, if the parent has any).

We get to hide managing this structure from the user.

* For aggregation, we walk backwards along the list of production slice indices — each child gets to produce output (all children must produce outputs of the same type), possibly sampling random choices. `SwitchCombinator` will provide the branching pattern here across grammar enumeration indices. The output gets captured in the aggregation sub-traces, meaning it’s amenable to incremental computation.

## JAX implementation

Here, I'll illustrate some JAX idioms which I expect to use in the implementation of this combinator.

**`jax.lax.scan` to support forward/reverse mode AD**

One way to express unbounded recursive computation (without dynamic allocation in each call) is `jax.lax.while_loop`. This operation would allow us to express a sort of "run until condition" pattern, and the implementation would require us to pre-allocate a choice map and then fill it -- and the generative computation would truncate either (a) when the choice map is filled or (b) when the condition fails.

However, `jax.lax.while_loop` is not compatible with reverse mode. So instead of using `jax.lax.while_loop`, I recommend using `jax.lax.scan`, which is reverse/fwd compatible. The difference here is that `jax.lax.scan` requires a statically known truncation limit.

For the GFI methods, we need to provide bounds for allocating for choice maps anyways. The downside to `jax.lax.scan` is that the operation will execute all the way up to the truncation limit each time. So there's a runtime trade off in return for reverse mode compatibility. To solve this for `Unfold` (for example), there's a switch statement inside the `jax.lax.scan` which either runs a step of the generative computation, or falls back on some default non-active values.

One fundamental fact here: attempting to avoid computation with switches is not possible, because GPUs will compute all branches of switch, and then `select` using the switch index.

So there's an interesting trade off here. One option is to allow the user to select what mode of implementation they want (e.g. select `jax.lax.while` -- you lose reverse mode differentiability) or -- even better -- we just use the mode which is appropriate **per GFI method**, so e.g. `assess`, which provides access to gradient functions, uses `jax.lax.scan` but all the others use `jax.lax.while`.

We should explore this implementation primitive switching strategy in another issue.

**Comment by sritchie** [comment]
Closing, migrated to internal Notion.
